More tests that exercise RPCs running across mixed versions

andrwng opened the pull request:
This PR adds tests that exercise RPC codepaths on clusters that span multiple versions. It leverages the recently introduced `MixedVersionWorloadRunner` to run workloads that deterministically sends RPCs in a specific direction, swapping direction and the binary versions of sender and receiver.

This includes all Raft traffic, transactions traffic, and most of traffic targeted at the controller leader.

Opened as a draft because it depends on https://github.com/redpanda-data/redpanda/pull/5282. Only the last four commits are new.

NyaliaLui commented:
@andrwng  Is this ready for review or are you still making edits?

NyaliaLui commented on tests/rptest/tests/upgrade_with_workload.py:
```diff
@@ -42,24 +45,32 @@ def upgrade_with_workload(redpanda: RedpandaService, initial_version,
         # Upgrade one node and send RPCs in both directions.
         installer.install([node0], RedpandaInstaller.HEAD)
-        redpanda.restart_nodes([node0])
+        redpanda.rolling_restart_nodes([node0],
```
Based on the commit message the true problem is that we need to improve health checks in the test infra instead of using rolling restart as a workaround. If that's true, then we should figure out those improvements instead.

I recognize that this is out of the scope of this PR, but maybe we should start a discussion over this.

andrwng replied:
Partially, though in general, waiting for the cluster to become healthy seems reasonable in the context of immediately running RPCs and workloads.

NyaliaLui reviewed (changes requested):
Looks like there were some problems in KafkaRpcTest

andrwng commented on tests/rptest/services/rolling_restarter.py:
```diff
@@ -28,6 +28,11 @@ def restart_nodes(self,
         the given configs.
         """
+        if not stop_timeout:
+            stop_timeout = 30
```
Yeah, I could pull these out into their own member variable, otherwise the existing node timeout seems a bit low. Will consider this

NyaliaLui reviewed (changes requested):
Looks like there's still an issue with KafkaRPC

Also, what other RPCs do we need to exercise? For example, partition balancer and controller for nodes joining the cluster.

andrwng commented:
Failed test was https://github.com/redpanda-data/redpanda/issues/5358

andrwng commented:
>Looks like there's still an issue with KafkaRPC
Fixed the Kafka RPC issue -- seems like upon restarting nodes, it's possible the rest of the cluster thinks a node A is leader, but it's been recently restarted and needs to establish leadership again before being able to properly service RPCs.

>Also, what other RPCs do we need to exercise? For example, partition balancer and controller for nodes joining the cluster.
Updated the CL with some more context.

NyaliaLui commented on tests/rptest/tests/upgrade_test.py:
```diff
@@ -311,3 +313,146 @@ def has_expected_topics():
+                                  partition=0)
+        # Avoid nodes of interest by operating on the back node.
+        back_node = self.redpanda.nodes[-1]
+        back_id = self.redpanda.idx(back_node)
```
I may be missing something here but does the MixedVersionWorkloadRunner never select the back node? What I'm confused by is why strictly selecting the last node is preferred over any others?

andrwng replied:
Yeah, the mixed version runner runs workloads between the first two nodes deterministically, exercising a specific direction of RPC traffic (e.g. from old version to new version, or vice versa). Added a comment.

NyaliaLui reviewed (commented):
The test structure looks OK to me. I left a few questions.

NyaliaLui commented:
Also, per our recent standup discussion, can you do a ci-repeat-5 on this this is a test for important subsystems such as raft.

andrwng commented:
/ci-repeat 10

andrwng commented on tests/rptest/tests/upgrade_test.py:
```diff
@@ -232,3 +236,78 @@ def raft_workload(self, src_node, dst_node):
+            dst_id = self.redpanda.idx(dst_node)
+            # Ensure there is an active controller leader (i.e. it's finished
+            # bootstrapping, etc) by creating a topic.
+            first_topic_name = rand_with_prefix("topic")
```
Ah thanks for the pointer! Done.

andrwng commented on tests/rptest/tests/upgrade_test.py:
```diff
@@ -232,3 +236,78 @@ def raft_workload(self, src_node, dst_node):
+
+
+def rand_with_prefix(prefix):
+    return f"{prefix}-{''.join([random.choice(string.ascii_letters) for _ in range(5)])}"
```
Done

andrwng commented on tests/rptest/tests/transactions_test.py:
```diff
@@ -244,3 +249,119 @@ def change_static_member_test(self):
+
+        def run_txn(should_commit):
+            self.generate_data(self.output_t, self.max_records)
+            producer = ck.Producer({
```
This is what's used in the rest of the test. Using `ck` also has the advantage in that it's fairly simple, it doesn't do too much in the background (e.g. no aggressive prefetching of leadership info).

andrwng commented:
It looks like the transaction test is failing now:

```
====================================================================================================
test_id:    rptest.tests.transactions_test.MixedVersionTransactionsTest.test_txn_rpcs_with_upgrade
status:     FAIL
run time:   2 minutes 10.358 seconds


    KafkaException(KafkaError{code=INVALID_TXN_STATE,val=48,str="Failed to initialize Producer ID: Broker: Producer attempted a transactional operation in an invalid state"})
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/ducktape/tests/runner_client.py", line 135, in run
    data = self.run_test()
  File "/usr/local/lib/python3.10/dist-packages/ducktape/tests/runner_client.py", line 227, in run_test
    return self.test_context.function(self.test)
  File "/root/tests/rptest/services/cluster.py", line 35, in wrapped
    r = f(self, *args, **kwargs)
  File "/root/tests/rptest/tests/transactions_test.py", line 573, in test_txn_rpcs_with_upgrade
    MixedVersionWorkloadRunner.upgrade_with_workload(
  File "/root/tests/rptest/tests/upgrade_with_workload.py", line 51, in upgrade_with_workload
    workload_fn(node0, node1)
  File "/root/tests/rptest/tests/transactions_test.py", line 551, in txn_workload
    run_txn(should_commit=True)
  File "/root/tests/rptest/tests/transactions_test.py", line 520, in run_txn
    producer.init_transactions()
cimpl.KafkaException: KafkaError{code=INVALID_TXN_STATE,val=48,str="Failed to initialize Producer ID: Broker: Producer attempted a transactional operation in an invalid state"}
```

This test passed last I ran it, so either I borked rebasing or something's been broken. @VadimPlh does anything look fishy in the transaction test to you? It goes through a partial upgrade, rollback, and full upgrade: https://github.com/redpanda-data/redpanda/blob/dev/tests/rptest/tests/upgrade_with_workload.py for reference.

VadimPlh commented:
> It looks like the transaction test is failing now:
> 
> ```
> ====================================================================================================
> test_id:    rptest.tests.transactions_test.MixedVersionTransactionsTest.test_txn_rpcs_with_upgrade
> status:     FAIL
> run time:   2 minutes 10.358 seconds
> 
> 
>     KafkaException(KafkaError{code=INVALID_TXN_STATE,val=48,str="Failed to initialize Producer ID: Broker: Producer attempted a transactional operation in an invalid state"})
> Traceback (most recent call last):
>   File "/usr/local/lib/python3.10/dist-packages/ducktape/tests/runner_client.py", line 135, in run
>     data = self.run_test()
>   File "/usr/local/lib/python3.10/dist-packages/ducktape/tests/runner_client.py", line 227, in run_test
>     return self.test_context.function(self.test)
>   File "/root/tests/rptest/services/cluster.py", line 35, in wrapped
>     r = f(self, *args, **kwargs)
>   File "/root/tests/rptest/tests/transactions_test.py", line 573, in test_txn_rpcs_with_upgrade
>     MixedVersionWorkloadRunner.upgrade_with_workload(
>   File "/root/tests/rptest/tests/upgrade_with_workload.py", line 51, in upgrade_with_workload
>     workload_fn(node0, node1)
>   File "/root/tests/rptest/tests/transactions_test.py", line 551, in txn_workload
>     run_txn(should_commit=True)
>   File "/root/tests/rptest/tests/transactions_test.py", line 520, in run_txn
>     producer.init_transactions()
> cimpl.KafkaException: KafkaError{code=INVALID_TXN_STATE,val=48,str="Failed to initialize Producer ID: Broker: Producer attempted a transactional operation in an invalid state"}
> ```
> 
> This test passed last I ran it, so either I borked rebasing or something's been broken. @VadimPlh does anything look fishy in the transaction test to you? It goes through a partial upgrade, rollback, and full upgrade: https://github.com/redpanda-data/redpanda/blob/dev/tests/rptest/tests/upgrade_with_workload.py for reference.

You used API which is not released. KIP-447 is not supported prev releases.

VadimPlh commented on tests/rptest/tests/transactions_test.py:
```diff
@@ -434,3 +444,132 @@ def get_tx_manager_broker():
+                    producer.produce(self.output_t.name, r.value(), r.key())
+                num_consumed_records += len(records)
+
+            producer.send_offsets_to_transaction(
```
This API does not supported old version. 
But also I think you do not needed it here

VadimPlh commented on tests/rptest/tests/transactions_test.py:
```diff
@@ -434,3 +444,132 @@ def get_tx_manager_broker():
+    @cluster(num_nodes=3)
+    def test_txn_rpcs_with_upgrade(self):
+        # Initialize a transaction to bootstrap our transaction internal topic.
+        producer = ck.Producer({
```
If producer got timeout for transaction it will be fatal error before upgrade. So you should recreate it in this case
